group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,454 @@
+        // in versions from v2 to v4. Otherwise, the retention defined on the broker is used. If an explicit
+        // commit timestamp is provided (v1 only), the expiration timestamp is computed based on that.
+        final OptionalLong expireTimestampMs = request.retentionTimeMs() == OffsetCommitRequest.DEFAULT_RETENTION_TIME ?
+            OptionalLong.empty() : OptionalLong.of(currentTimeMs + request.retentionTimeMs());
+        final OffsetCommitResponseData response = new OffsetCommitResponseData();
+        final List<Record> records = new ArrayList<>();
+
+        request.topics().forEach(topic -> {
+            final OffsetCommitResponseData.OffsetCommitResponseTopic topicResponse =
+                new OffsetCommitResponseData.OffsetCommitResponseTopic()
+                    .setName(topic.name());
+            response.topics().add(topicResponse);
+
+            topic.partitions().forEach(partition -> {
+                if (partition.committedMetadata() != null && partition.committedMetadata().length() > offsetMetadataMaxSize) {
+                    topicResponse.partitions().add(new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.OFFSET_METADATA_TOO_LARGE.code()));
+                } else {
+                    log.debug("[GroupId {}] Committing offsets {} for partition {}-{} from member {} with leader epoch {}.",
+                        request.groupId(), partition.committedOffset(), topic.name(), partition.partitionIndex(),
+                        request.memberId(), partition.committedLeaderEpoch());
+
+                    topicResponse.partitions().add(new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.NONE.code()));
+
+                    final OptionalInt leaderEpoch = partition.committedLeaderEpoch() == RecordBatch.NO_PARTITION_LEADER_EPOCH ?
+                        OptionalInt.empty() : OptionalInt.of(partition.committedLeaderEpoch());
+                    final String metadata = partition.committedMetadata() == null ?
+                        OffsetAndMetadata.NO_METADATA : partition.committedMetadata();
+                    final long commitTimestampMs = partition.commitTimestamp() == OffsetCommitRequest.DEFAULT_TIMESTAMP ?
+                        currentTimeMs : partition.commitTimestamp();
+
+                    final OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(
+                        partition.committedOffset(),
+                        leaderEpoch,
+                        metadata,
+                        commitTimestampMs,
+                        expireTimestampMs
+                    );
+
+                    records.add(RecordHelpers.newOffsetCommitRecord(
```

Comment:
do we no longer care about MemoryRecords.builder() because we append OffsetCommitRecords directly?

will we be adding the compression type in a following PR?

Reply:
The MemoryRecords is built in the `CoordinatorPartitionWriter`, isn't it? The compression is also applied there.

Reply:
ah right. thanks for the clarification.